content/posts/2023-12-optimising-spark-memory-usage.md

```diff
@@ -0,0 +1,278 @@
+
+```js
+const measurements = /* fetch and parse JSON */
+const validMeasurements = measurements.map(measurement => {
+  try {
+    return new Measurement(measurement)
+  } catch (err) {
+    logger.error('Invalid measurement:', err.message, measurement)
+    return null
+  }
+})
+```
+
+Take a look at the pull request
+[filecoin-station/spark-evaluate#84](https://github.com/filecoin-station/spark-evaluate/pull/84) if
+you are interested in more details.
+
+## Duplicate string values
+
+This first improvement sparked my interest. What other easy optimizations we can make?
+
+Let's take a look at an example measurement:
+
+```json
+{
+  "id": 280847321,
+  "spark_version": "1.6.0",
+  "zinnia_version": "0.15.0",
+  "participant_address": "0x000000000000000000000000000000000000dEaD",
+  "finished_at": "2023-12-11T08:31:00.157Z",
+  "timeout": false,
+  "start_at": "2023-12-11T08:30:29.947Z",
+  "status_code": 502,
+  "first_byte_at": null,
+  "end_at": "2023-12-11T08:30:59.954Z",
+  "byte_length": 0,
+  "attestation": null,
+  "inet_group": "uWJAEcUyzouo",
+  "car_too_large": false,
+  "cid": "bafybeice7venemgas2vijqc6xfdf6zequrxziajapm2j6hxckeu6fimjwa",
+  "provider_address": "/ip4/207.189.112.60/tcp/26197/http",
+  "protocol": "http"
+}
+```
+
+Setting aside timestamps, the following properties store a string value:
+
+- `spark_version` contains the version of the SPARK node that performed the retrieval.
+- `zinnia_version` is a field we ignore now; we don't keep it in `Measurement` instances.
+- `participant_address` contains the Station's wallet address in ETH format.
+- `inet_group` is an anonymous id of the IPv4 /24 subnet from which the measurement was submitted.
+- `cid` is the identifier of the content that the node attempted to retrieve.
+- `provider_address` is the address of the Storage Provider.
+- `protocol` is either `http`, `bitswap`, or `graphsync`.
+
+There is only a small number of possible values; each value is repeated many times in different
+measurements. For example, we pick 4000 different CIDs to test every round. With 100k measurements
+submitted per round, there are, on average, 10,000 measurements for the same CID.
+
+Here comes the question: how does V8 (the JavaScript runtime in Node.js) represent strings? If two
+object properties hold the same string value, how many copies of the string's bytes are stored in
+the memory? The answer seems to be "two copies".
+
+Aside: I guess this makes sense - most applications don't have that many duplicate string values.
+The cost of maintaining a lookup table would outweigh the benefits.
+
+Using the knowledge about the specific usage patterns of our application, we can trade additional
+CPU load for less memory usage and implement a lookup table ourselves. The idea is simple: whenever
+we encounter a new string coming from the parsed JSON data, check if we have seen it before. If yes,
+replace the new string value with a pointer to the same string we have seen before.
+
```

Comment:
Idea: add a diagram showing many measurement objects where the same CIDs point to the same string instance.